resources/assets/js/components/Modal.tsx

```diff
@@ -1,189 +1,168 @@
-import React, { Component, createRef } from "react";
-
-export interface ModalProps {
-  /** HTML ID for modal attributes */
-  id: string;
-  /** Title that appears within the modal when open */
-  title: string;
-  /** Optional subtitle that appears within the open modal */
-  subtitle?: string;
-  /** Text displayed on the button that displays the modal */
-  openText: string;
-  /** Text displayed on the button that hides the modal */
-  closeText: string;
-  /** Text displayed on the modal confirmation button */
-  confirmText: string;
-  /** Text displayed on the modal cancellation button */
-  cancelText: string;
-  /** React children */
-  children?: React.ReactNode;
-  /** Function to run for modal confirmation click */
-  handleConfirm: (event: React.MouseEvent<HTMLButtonElement>) => void;
-}
-
-interface ModalState {
-  /** Height of the modal itself */
-  height: number;
-  /** Visibility of the modal */
-  visible: boolean;
-}
-
-class Modal extends Component<ModalProps, ModalState> {
-  private divElement = createRef<HTMLDivElement>();
-
-  public constructor(props: ModalProps) {
-    super(props);
-
-    this.state = { height: 0, visible: false };
-  }
-
-  public componentDidMount = (): void => {
-    const node = this.divElement.current;
-    if (node) {
-      const height = node.clientHeight;
-      this.setState({ height });
+import React, {
+  createContext,
+  useContext,
+  useEffect,
+  useRef,
+  useState,
+} from "react";
+import { createPortal } from "react-dom";
+
+const modalContext = createContext({
+  id: "",
+  parentElement: "",
+  visible: false,
+  onModalConfirm: undefined,
+  onModalCancel: undefined,
+});
+
+export default function Modal({
+  id,
+  parentElement,
+  visible,
+  children,
+  onModalConfirm,
+  onModalCancel,
+}): React.ReactPortal {
+  // Set up div ref to measure modal height
+  const modalRef = useRef<HTMLDivElement>(null);
+
+  // Internal state to keep track of the overflow setting
+  const [overflow, setOverflow] = useState("");
+
+  const handleTabKey = (e: KeyboardEvent): void => {
+    if (modalRef && modalRef.current) {
+      const focusableModalElements = modalRef.current.querySelectorAll(
+        'a[href], button, textarea, input[type="text"], input[type="email"], input[type="radio"], select',
+      );
+      const firstElement = focusableModalElements[0] as HTMLElement;
+      const lastElement = focusableModalElements[
+        focusableModalElements.length - 1
+      ] as HTMLElement;
+
+      if (!e.shiftKey && document.activeElement !== firstElement) {
```

Comment:
If I understand correctly, this will always jump between the first and last element. That doesn't work if you add more focusable elements in the modal body.

Reply:
This just handles the cases where we need to "trap" focus in the modal. Outside of the first or last element in the map, it just tabs through normally.

Reply:
Nope, I'm wrong. Working on refactoring into something that works properly.